### ISSUE TITLE
explicitly pass in names for delayed functions

### ISSUE THREAD

**Issue by maxhutch** [dask/dask#1239]
tl;dr I want to pass the uniquely identifying name of delayed function instead of having it automatically generated via a hash of the function and arguments.

I'm using delayed to build workflows.  The result of a function that I'm delaying depends on some but not all of its arguments.  For example, a function which performs a parallel reduction might have a `num_workers` argument that doesn't affect the output.  Right now, delayed hashes it anyways and forces re-computation.  I want to avoid that.

The real use-case is a bit more complicated: I'm making some system calls from within the delayed functions that run simulations on supercomputers.  This calculations are very expensive, so I want a bit more control over which tasks are considered to have the same output.

The behavior I'm looking for [is here](https://github.com/maxhutch/nekpy/blob/names/nekpy/dask/tasks.py#L60), but it doesn't work: the names in the dask are still of the form `<function name>-<hash of something>`.  Tracing the code I get lost in DelayedLeaf, which seems to ignore its name argument.

**Comment by jcrist** [comment]
You can provide the name yourself explicitly using the `dask_key_name` kwarg:

```python
In [105]: foo
Out[105]: Delayed('fooeda225def65bca4b5be25f9f12f1bbaa')

In [106]: foo(1, 2, 3)
Out[106]: Delayed('foo-2b595b88120141ef301c7ccab85a386b')

In [107]: foo(1, 2, 3, dask_key_name='foo')
Out[107]: Delayed('foo')
```

Is that an acceptable solution here?

**Comment by maxhutch** [comment]
When does that kwarg need to be passed?  How does that differ from the intended behavior of the `name` argument to `delayed`?

**Comment by jcrist** [comment]
The `name` kwarg sets the name in the underlying graph for the object being wrapped. The `dask_key_name` kwarg sets the key name for the result of a `__call__` on a delayed object. The following might clarify things:

```python
In [112]: def foo(a, b, c=0):
   .....:     return a + b + c
   .....:

In [113]: dfoo = delayed(foo, name='foo', pure=True)

In [114]: dfoo
Out[114]: Delayed('foo')

In [115]: dfoo(1, 2, c=3)
Out[115]: Delayed('foo-7012413eafb786ebcdd430865623469c')

In [116]: dfoo(1, 2, c=3, dask_key_name='foo_called')
Out[116]: Delayed('foo_called')
```

**Comment by maxhutch** [comment]
Yeah, that works and makes sense.  I think its the currying that threw me off.  
Could `dask_key_name` get promoted to the docs?  This is really helpful functionality.

**Comment by jcrist** [comment]
Good point. Adding.
